README.md
Kafe (`/ka'fe:/`) is a minimal, portable, open-source remote automation framework, task runner,
and build automation tool.

It is used as a common task runner, for local and remote automation, application deployment,
build system development, build pipeline management, and more!

## Installation

See [available package downloads and repositories](DOWNLOAD.md).

Kafe officially supports:

- **Enterprise Linux** versions [7<sup>1</sup>](./DOWNLOAD.md#el7), [8](./DOWNLOAD.md#el8)
- **Fedora** versions [31](./DOWNLOAD.md#fedora-31), [32](./DOWNLOAD.md#fedora-32), [33](./DOWNLOAD.md#fedora-33),
  [34](./DOWNLOAD.md#fedora-34), [35](./DOWNLOAD.md#fedora-35)
- **Ubuntu** versions [18.04](./DOWNLOAD.md#ubuntu-1804), [20.04](./DOWNLOAD.md#ubuntu-2004),
  [20.10](./DOWNLOAD.md#ubuntu-2010), [21.04](./DOWNLOAD.md#ubuntu-2104)
- **Debian** versions [9](./DOWNLOAD.md#debian-9), [10](./DOWNLOAD.md#debian-10), [11](./DOWNLOAD.md#debian-11)
- **macOS** versions 10.5 (Catalina), 11 (Big Sur) and newer,
  using [Homebrew vendor tap](https://github.com/libkafe/homebrew-kafe). Both x86 and Apple Silicon is supported.

<sup>1</sup> - EL7 does not ship with Lua 5.3. We currently use
[Cheese](http://www.nosuchhost.net/~cheese/fedora/packages/epel-7/x86_64/cheese-release.html) repository to obtain Lua
5.3 on EL7.

You can also install Kafe by building it from source code locally using `CMake` - provided all dependencies are met,
you should be able to build Kafe on nearly any Linux and UNIX operating system. See below for instructions on how
to build Kafe from the source code.

## Writing Kafe scripts

You can declare your server inventory and automation tasks in Kafe scripts. Kafe scripts are written in
[Lua](https://www.lua.org) programming language. Kafe CLI looks for a file named `kafe.lua`
in current working directory to execute the tasks from.

### Why Lua?

Lua is widely used, minimal and easy to learn programming language. It is easy to adopt and has
[extensive documentation](http://www.lua.org/docs.html) with large and stable standard library - a perfect fit for
scripting against Kafe APIs.

### Main concepts

#### Tasks

All automation code in Kafe is divided in work units called tasks. A task is isolated unit of work that describes all
actions needed to perform against local or remote system.

Tasks are independent, and there are no facilities to chain or relate tasks - this is by design. To share code and
operations between tasks, you can use standard Lua functions.

#### Inventory

Kafe categorizes all remote servers by environment and role. An environment is a logical unit of remote servers -
testing environment, staging environment, production environment, and so forth. A role is a logical unit of one or more
servers performing the same infrastructure role - a database server, a web server, an application host, to name a few.
Each inventory item can have one or more roles and environment.

All remote servers in Kafe must be placed in inventory. Inventory is a list of remote servers with associated roles and
environments. Each server can have one or more roles and can be present in one or more environments. Each inventory line
uniquely combines remote server address, one role, and one environment. To add the same remote server to multiple roles
and environments duplicate the inventory definition line.

#### Execution of tasks

All tasks are executed in order as invoked by the script (top-down). This is by design. Parallelization of remote
management tasks might seem a good idea at first, but it rarely is - parallelization makes remote error handling much
more difficult, might leave remote systems in an inconsistent state and requires a much larger investment in writing the
automation tasks.

### Example

Example bellow demonstrates a very basic usage of Kafe. Kafe is designed to be purpose agnostic - you can use it for any
kind of local and remote automation tasks, running shell commands, uploading files, etc.

```lua
-- Include the API
local k = require('kafe')

-- Make sure our deployment script is run with compatible libkafe
k.require_api(1)

-- The username to use for SSH connection
local username = os.getenv('USER');

-- Add a server to inventory [username, hostname, port, environment, role]
k.add_inventory(username, 'example.com', 22, 'production', 'app')

-- This is a remote task
-- This task can be invoked like: `kafe do production whoami`
-- Remote tasks must have subtasks
k.task('whoami', function()
    -- Commands here are executed on localhost

    -- Execute a subtask on inventory node with role 'app'
    k.on('app', function()
        -- Commands here are executed on a remote server
        k.shell('whoami')
    end)
end)

-- This is a local task. Local tasks do not need to have subtasks.
-- This task can be invoked like: `kafe local date`
k.task('date', function()
    k.shell('date')
end)
```

#### More examples

You can find more examples on how to write Kafe scripts in the [examples](./examples) directory.

### Scripting API

Scripting API documentation is available [here](docs/SCRIPTING_API_L1.md).

### Using the CLI

To execute a Kafe script, you need to invoke `kafe` command with sub-command `do`. You need to pass the desired
execution environment and a list of one or more tasks to be executed, separated by a comma.

For example, to execute example tasks named `deploy`, `verify` on `production` environment, you would execute following
Kafe CLI command:

`kafe do production deploy,verify`

When executed, Kafe CLI will look for a file named `kafe.lua` in the current working direcory. This file will be
interpreted and requested tasks from it will be executed against all relevant remote servers.

### Local tasks and local execution environment

From version 1.1.4, Kafe adds support for local scripting using `kafe local` command line interface. This is simply a
shortcut that allows for writing local automation tasks, local scripts and similar directly in Kafe and allows Kafe to
be used as build automation framework and local task runner.

**NOTE:** in local, APIs like `exec`, `shell` are simply aliased to `local_shell` and `local_exec`.

#### SSH and SSH authentication

Kafe is using SSH for remote command execution and file uploads. It will attempt to use SSH agent, any known local SSH
keys, as well as GSSAPI-WITH-MIC and password based authentication to authenticate to remote hosts.

If no SSH agent is present, you can set the passphrase to be used to decrypt any encrypted private keys using
environment variable `KAFE_SSH_PKEY_PASS`. You can set password to be used for password based authentication using
environment variable named `KAFE_SSH_USER_PASS`.

**IMPORTANT:** Kafe will not automatically add remote keys to known hosts nor will it provide a way to do so
interactively. It is your responsibility to ensure remote host keys are added to known hosts before attempting to
connect to remote hosts using Kafe. Any attempts to connect to remote hosts with unknown or changed host keys will fail.

### Debugging

You can change the logging level of the CLI tool by setting `KAFE_LOG_LEVEL` environment variable. For example:

`KAFE_LOG_LEVEL=1 kafe do staging deploy`

Available log levels are:

ALL = 0, TRACE = 1, DEBUG = 2, INFO = 3, SUCCESS = 4, WARNING = 5, ERROR = 6, NONE = 7.

## Kafe and libkafe

Kafe is written entirely in C++ - it is distributed as a native binary (kafe-cli) and a shared library (libkafe)
with C++ development headers made available for embedding (libkafe-dev).

## Operating system support

Kafe should work with any UNIX-like operating system, provided all external dependencies are met.

Windows is not supported and there are no plans to support Windows builds. Windows users should be able to use Windows
Subsystem for Linux to run or build Kafe. Take note that contributions and issues related to support for Windows will
not be accepted.

macOS is officially supported starting from version 10.15 Catalina. It is entirely possible Kafe works just fine on
older versions of macOS, but no testing has been done. If you are unable to upgrade to latest macOS version, you should
still try to install Kafe from sources. It will most likely work just fine provided all external dependencies are met.

If you would like to contribute your distribution to the above list, you can implement a build for it. Just follow the
samples already present in [dist](./dist) folder. Any contributions adding support for new distributions must contain
all current non-EOL distribution versions, including RC versions, if any.

#### Is Kafe available on Homebrew?

Yes, you can install Kafe from [lbikafe tap](https://github.com/libkafe/homebrew-kafe).

#### Does Kafe have official Docker images with binaries?

No and it is unlikely there ever will be. At this time, Kafe is primarily designed to be used as a CLI tool installed on
a local system - having a Docker image for it has as much benefit as having a Docker image for `tail`. If you need to
have Kafe in Docker for CI or some other use, creating your own image using any of the supported base operating system
images should be trivial.

## Backwards compatibility

### Kafe CLI and libkafe

Kafe CLI should be compatible with libkafe within the same major version.

### libkafe

Libkafe development libraries are guaranteed to be compatible within the same major version.

### Scripting API

Scripting API is versioned separately from CLI and libkafe. Scripting API compatibility is guaranteed within the same
API version, meaning, any script built for API version 1 will work with any Kafe version supporting version 1 regardless
of the internals of Kafe itself.

| libkafe version | API level    | Lua      | Status     |
|-----------------|--------------|----------|------------|
| 1.x             | 1            | 5.3, 5.4 | Locked     |
| 2.x             | 2<sup>1</sup>| 5.3, 5.4 | Future     |

<sup>1</sup> - it is very likely API level 2 will be fully backwards compatible with level 1.

### Lua compatibility

Kafe currently supports Lua version 5.3 with 5.4 support in development. Lua versions older than 5.3 are not supported.
Kafe will build against the latest available version of Lua in the build system. For the most part, it is the latest
version distributed by the operating system itself with the sole exception of CentOS 7, where third party repository is
required to install and use Kafe.

### Building from source

To build Kafe from the source, you will need C++ and C compiler, CMake (version 3.11.4 or newer), Make and following
development libraries:

- liblua version 5.3 or newer (up to version 5.4)
- libcurl (reserved for future APIs)
- libarchive
- libssh
- libgit2 (reserved for future APIs)

Kafe is built and tested using Clang toolchain - version 7 or higher, depending on build environment, both in
development and distribution. You should be able to build your own binaries using any C11 and C++17 compatible C/C++
compiler toolchain.

See [/dist/](./dist) directory for example Docker build files and scripts per each supported build target.

You can also refer to [build-dist.sh](./build-dist.sh) for how to build distribution specific packages using Docker.

#### Building on macOS

To build from sources on macOS you will need `git` and Kafe dependencies installed on your system.

To install dependencies using Homebrew, use this command:

`brew install llvm lua libssh libgit2 curl libarchive`

To build

1. Clone the source from `git@github.com:libkafe/kafe.git`;
2. Optionally, check out the desired version tag to build from GIT;
3. Execute [build-dist-macos.sh](./build-dist-macos.sh) file found at the root of the cloned sources.
4. Optionally, enter `build/osx` directory, and issue `make install` command to install Kafe on your local machine.

## Reporting issues

To report a bug or request a feature, open a ticket
at [https://github.com/libkafe/kafe/issues](https://github.com/libkafe/kafe/issues).

For security related issues, please follow our [Vulnerability Disclosure Policy](./SECURITY.md).

## History

Kafe is nearly complete rewrite of OPM (Optional Package Manager) - a command-line tool I wrote several years ago for my
personal use to automate application and content deployment. It was written entirely in C using Python as a scripting
language to interface against remote APIs. OPMs design had several drawbacks, namely very tight integration with
specific Python version and gratuitous use of static linking against certain 3rd party dependencies. It worked (and
still does!)
just fine, but was not suitable at all to be open sourced nor was it easily sharable - meaning it was fairly useless
outside the realm of my own personal use.

[Hemp](https://github.com/Addvilz/hemp) was another tool I wrote prior to Kafe for use in production to automate
deployment and remote automation. This tool was based on now unmaintained 3rd party remote automation library and was
deprecated due to lack of support for Python 3 and breaking changes from upstream library vendor. Unfortunately, Hemp
was and still is being used in a number of production projects and many production systems rely on it being functional
and maintained, something that is no longer viable.

I created Kafe to replace both of these tools. Kafe is designed to be as minimal and straightforward as possible. I
wanted to create an automation tool with strong backwards compatibility guarantees, meaning that once written,
automation tasks should work with minimal or no changes years to come.

## License

Licensed under the Apache License, Version 2.0 (the "License"). See [LICENSE](./LICENSE) for more details.


CHANGELOG
* Sun Jul 18 2021 Matiss Treinis <matiss@hub256.com> - 1.2.0
- Introduce make command and build pipelines

* Mon May 24 2021 Matiss Treinis <mrtreinis@gmail.com> - 1.1.5.1
- Support for Apple SoC

* Mon May 24 2021 Matiss Treinis <mrtreinis@gmail.com> - 1.1.5
- Bugfix - within aliased to shell in local context.

* Mon May 24 2021 Matiss Treinis <mrtreinis@gmail.com> - 1.1.4
- Add "local" command for local execution context runner
- Introduce strict mode
- Switch EL 8 compatible build to use AlmaLinux 8
- Add support for Fedora 35
- Remove support for Ubuntu 19.10 (EOL)

* Mon Oct 5 2020 Matiss Treinis <mrtreinis@gmail.com> - 1.1.3
- Fix build compatibility with libssh
- Fix OSX build

* Thu Aug 6 2020 Matiss Treinis <mrtreinis@gmail.com> - 1.1.2
- Add support for .kafeignore files.

* Thu Jun 25 2020 Matiss Treinis <mrtreinis@gmail.com> - 1.1.1
- Update build configuration, define SOVERSION. No functional changes.

* Wed Mar 11 2020 Matiss Treinis <mrtreinis@gmail.com> - 1.1.0
- Logging timer no longer reports nonsense values (#1).
- Environment name is now available from within scripts (#2).
- Archives created using the API now have proper timestamps and metadata (#3).
- Empty error messages are no longer possible in libkafe to avoid ambiguous errors (#4).
- Extra arguments to "do" command are now forwarded to task function (#5).
- SSH key passphrase and user password can now be defined in executing environment (#6).

* Wed Mar 4 2020 Matiss Treinis <mrtreinis@gmail.com> - 1.0.0
- Initial public release


build-dist.sh
#!/usr/bin/env bash
set -xe

# TODO - I do not like this copy paste...

rm -rf build/
rm -rf build-artifact/

# EL7
if [[ "$(docker images -q "kafe/centos:7-build" 2> /dev/null)" == "" ]]; then
  docker build -t kafe/centos:7-build dist/centos/7
fi

docker run -it --rm -v `pwd`:/kafe kafe/centos:7-build
mkdir -p build-artifact/centos-7/
cp build/centos/7/kafe-cli-*.rpm build-artifact/centos-7/
cp build/centos/7/libkafe-*.rpm build-artifact/centos-7/
for f in build-artifact/centos-7/*; do mv -v "$f" $(echo "$f" | sed "s/\.rpm/\.el7\.rpm/"); done

# EL 8
if [[ "$(docker images -q "kafe/almalinux:8-build" 2> /dev/null)" == "" ]]; then
  docker build -t kafe/almalinux:8-build dist/almalinux/8
fi

docker run -it --rm -v `pwd`:/kafe kafe/almalinux:8-build
mkdir -p build-artifact/almalinux-8/
cp build/almalinux/8/kafe-cli-*.rpm build-artifact/almalinux-8/
cp build/almalinux/8/libkafe-*.rpm build-artifact/almalinux-8/
for f in build-artifact/almalinux-8/*; do mv -v "$f" $(echo "$f" | sed "s/\.rpm/\.el8\.rpm/"); done

# Debian 9
if [[ "$(docker images -q "kafe/debian:9-build" 2> /dev/null)" == "" ]]; then
  docker build -t kafe/debian:9-build dist/debian/9
fi

docker run -it --rm -v `pwd`:/kafe kafe/debian:9-build
mkdir -p build-artifact/debian-9/
cp build/debian/9/kafe-cli*.deb build-artifact/debian-9/
cp build/debian/9/libkafe*.deb build-artifact/debian-9/
for f in build-artifact/debian-9/*; do mv -v "$f" $(echo "$f" | sed "s/\.deb/\.deb9\.deb/"); done

# Debian 10
if [[ "$(docker images -q "kafe/debian:10-build" 2> /dev/null)" == "" ]]; then
  docker build -t kafe/debian:10-build dist/debian/10
fi

docker run -it --rm -v `pwd`:/kafe kafe/debian:10-build
mkdir -p build-artifact/debian-10/
cp build/debian/10/kafe-cli*.deb build-artifact/debian-10/
cp build/debian/10/libkafe*.deb build-artifact/debian-10/
for f in build-artifact/debian-10/*; do mv -v "$f" $(echo "$f" | sed "s/\.deb/\.deb10\.deb/"); done

# Debian 11
if [[ "$(docker images -q "kafe/debian:11-build" 2> /dev/null)" == "" ]]; then
  docker build -t kafe/debian:11-build dist/debian/11
fi

docker run -it --rm -v `pwd`:/kafe kafe/debian:11-build
mkdir -p build-artifact/debian-11/
cp build/debian/11/kafe-cli*.deb build-artifact/debian-11/
cp build/debian/11/libkafe*.deb build-artifact/debian-11/
for f in build-artifact/debian-11/*; do mv -v "$f" $(echo "$f" | sed "s/\.deb/\.deb11\.deb/"); done

# Ubuntu 18.04
if [[ "$(docker images -q "kafe/ubuntu:1804-build" 2> /dev/null)" == "" ]]; then
  docker build -t kafe/ubuntu:1804-build dist/ubuntu/1804
fi

docker run -it --rm -v `pwd`:/kafe kafe/ubuntu:1804-build
mkdir -p build-artifact/ubuntu-1804/
cp build/ubuntu/1804/kafe-cli*.deb build-artifact/ubuntu-1804/
cp build/ubuntu/1804/libkafe*.deb build-artifact/ubuntu-1804/
for f in build-artifact/ubuntu-1804/*; do mv -v "$f" $(echo "$f" | sed "s/\.deb/\.ubu1804\.deb/"); done

# Ubuntu 20.04
if [[ "$(docker images -q "kafe/ubuntu:2004-build" 2> /dev/null)" == "" ]]; then
  docker build -t kafe/ubuntu:2004-build dist/ubuntu/2004
fi

docker run -it --rm -v `pwd`:/kafe kafe/ubuntu:2004-build
mkdir -p build-artifact/ubuntu-2004/
cp build/ubuntu/2004/kafe-cli*.deb build-artifact/ubuntu-2004/
cp build/ubuntu/2004/libkafe*.deb build-artifact/ubuntu-2004/
for f in build-artifact/ubuntu-2004/*; do mv -v "$f" $(echo "$f" | sed "s/\.deb/\.ubu2004\.deb/"); done

# Ubuntu 20.10
if [[ "$(docker images -q "kafe/ubuntu:2010-build" 2> /dev/null)" == "" ]]; then
  docker build -t kafe/ubuntu:2010-build dist/ubuntu/2010
fi

docker run -it --rm -v `pwd`:/kafe kafe/ubuntu:2010-build
mkdir -p build-artifact/ubuntu-2010/
cp build/ubuntu/2010/kafe-cli*.deb build-artifact/ubuntu-2010/
cp build/ubuntu/2010/libkafe*.deb build-artifact/ubuntu-2010/
for f in build-artifact/ubuntu-2010/*; do mv -v "$f" $(echo "$f" | sed "s/\.deb/\.ubu2010\.deb/"); done

# Ubuntu 21.04
if [[ "$(docker images -q "kafe/ubuntu:2104-build" 2> /dev/null)" == "" ]]; then
  docker build -t kafe/ubuntu:2104-build dist/ubuntu/2104
fi

docker run -it --rm -v `pwd`:/kafe kafe/ubuntu:2104-build
mkdir -p build-artifact/ubuntu-2104/
cp build/ubuntu/2104/kafe-cli*.deb build-artifact/ubuntu-2104/
cp build/ubuntu/2104/libkafe*.deb build-artifact/ubuntu-2104/
for f in build-artifact/ubuntu-2104/*; do mv -v "$f" $(echo "$f" | sed "s/\.deb/\.ubu2104\.deb/"); done

# Fedora 31
if [[ "$(docker images -q "kafe/fedora:31-build" 2> /dev/null)" == "" ]]; then
  docker build -t kafe/fedora:31-build dist/fedora/31
fi

docker run -it --rm -v `pwd`:/kafe kafe/fedora:31-build
mkdir -p build-artifact/fedora-31/
cp build/fedora/31/kafe-cli-*.rpm build-artifact/fedora-31/
cp build/fedora/31/libkafe-*.rpm build-artifact/fedora-31/
for f in build-artifact/fedora-31/*; do mv -v "$f" $(echo "$f" | sed "s/\.rpm/\.f31\.rpm/"); done

# Fedora 32
if [[ "$(docker images -q "kafe/fedora:32-build" 2> /dev/null)" == "" ]]; then
  docker build -t kafe/fedora:32-build dist/fedora/32
fi

docker run -it --rm -v `pwd`:/kafe kafe/fedora:32-build
mkdir -p build-artifact/fedora-32/
cp build/fedora/32/kafe-cli-*.rpm build-artifact/fedora-32/
cp build/fedora/32/libkafe-*.rpm build-artifact/fedora-32/
for f in build-artifact/fedora-32/*; do mv -v "$f" $(echo "$f" | sed "s/\.rpm/\.f32\.rpm/"); done

# Fedora 33
if [[ "$(docker images -q "kafe/fedora:33-build" 2> /dev/null)" == "" ]]; then
  docker build -t kafe/fedora:33-build dist/fedora/33
fi

docker run -it --rm -v `pwd`:/kafe kafe/fedora:33-build
mkdir -p build-artifact/fedora-33/
cp build/fedora/33/kafe-cli-*.rpm build-artifact/fedora-33/
cp build/fedora/33/libkafe-*.rpm build-artifact/fedora-33/
for f in build-artifact/fedora-33/*; do mv -v "$f" $(echo "$f" | sed "s/\.rpm/\.f33\.rpm/"); done

# Fedora 34
if [[ "$(docker images -q "kafe/fedora:34-build" 2> /dev/null)" == "" ]]; then
  docker build -t kafe/fedora:34-build dist/fedora/34
fi

docker run -it --rm -v `pwd`:/kafe kafe/fedora:34-build
mkdir -p build-artifact/fedora-34/
cp build/fedora/34/kafe-cli-*.rpm build-artifact/fedora-34/
cp build/fedora/34/libkafe-*.rpm build-artifact/fedora-34/
for f in build-artifact/fedora-34/*; do mv -v "$f" $(echo "$f" | sed "s/\.rpm/\.f34\.rpm/"); done

# Fedora 35
if [[ "$(docker images -q "kafe/fedora:35-build" 2> /dev/null)" == "" ]]; then
  docker build -t kafe/fedora:35-build dist/fedora/35
fi

docker run -it --rm -v `pwd`:/kafe kafe/fedora:35-build
mkdir -p build-artifact/fedora-35/
cp build/fedora/35/kafe-cli-*.rpm build-artifact/fedora-35/
cp build/fedora/35/libkafe-*.rpm build-artifact/fedora-35/
for f in build-artifact/fedora-35/*; do mv -v "$f" $(echo "$f" | sed "s/\.rpm/\.f35\.rpm/"); done
